### PR TITLE
Extend the CLI to also call API methods

### PR DESCRIPTION
**Pull request by quantumagi**
This PR:
- Extends the CLI to also call API methods
- Adds the API feature to StratisD

### CODE REVIEW

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 1]
```diff
@@ -10,20 +14,31 @@ namespace Stratis.Bitcoin.Cli
...
             {
-                if (NodeSettings.PrintHelp(args, Network.Main))
+                if (args.Length == 0 || args.Contains("-help") || args.Contains("--help"))
                 {
-                    // TODO: add more outputs
-                    // Proposal: use reflection to find all settings classes and print out help.
-                    RpcSettings.PrintHelp(Network.Main);
+                    var builder = new StringBuilder();
+                    builder.AppendLine("Usage:");
+                    // TODO: Shouldn't this be dotnet run instead of dotnet exec?
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Cli/Program.cs]: yes, it should be `dotnet run`.

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 2]
```diff
@@ -35,22 +50,65 @@ public static void Main(string[] args)
...
+                    using (var client = new HttpClient())
+                    {
+                        client.DefaultRequestHeaders.Accept.Clear();
+                        client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+                        var url = $"{ApiURI}api/{args.ElementAt(1)}?{string.Join("&", args.Skip(2))}";
+                        try
+                        {
+                            var response = client.GetStringAsync(url).GetAwaiter().GetResult();
+                            // Attempt to format the response
+                            try
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Cli/Program.cs]: Do we need this `try` and the following `finally`?

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 3]
```diff
@@ -35,22 +50,65 @@ public static void Main(string[] args)
...
+                        Console.WriteLine(err.Message);
+                    }
+                }
+                else
+                {
+                    // Process API call
 
-                // Execute the RPC command
-                RPCResponse response = rpc.SendCommand(args.ElementAt(1), args.Skip(2).ToArray());
+                    var ApiURI = "http://localhost:37220/";
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Cli/Program.cs]: 37220 is for Bitcoin, and 37221 is for Stratis.

File: src/Stratis.Bitcoin.Cli/Program.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Cli/Program.cs]: Changed.

File: src/Stratis.Bitcoin.Cli/Program.cs (hunk 3 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Cli/Program.cs]: Added.

File: src/Stratis.Bitcoin.Cli/Program.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Cli/Program.cs]: You are right. All return values are JSON and errors follow a different path. Removed.

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 4]
```diff
@@ -10,53 +14,136 @@ namespace Stratis.Bitcoin.Cli
...
-                    RpcSettings.PrintHelp(Network.Main);
+                    var builder = new StringBuilder();
+                    builder.AppendLine("Usage:");
+                    builder.AppendLine(" dotnet run <Stratis.Bitcoin.Cli/Stratis.Bitcoin.Cli.dll> [network-name] [command] [arguments]");
+                    builder.AppendLine();
+                    builder.AppendLine("Command line arguments:");
+                    builder.AppendLine();
+                    builder.AppendLine($"[network-name]            Name of the network - e.g. \"stratis\".");
+                    builder.AppendLine($"[command]                 Name of RPC method or API <controller>/<method>.");
+                    builder.AppendLine($"[arguments]               Argument by position (RPC) or Name = Value pairs (API).");
```
Comment (Aprogiena) [inline src/Stratis.Bitcoin.Cli/Program.cs]: why the three lines with interpolated strings here?

and why to use string builder at all here and not just Console.WriteLine line by line?

if you cared about writing all at once (which i don't think you do), you could do it like this anyway:
```
Console.WrilteLine("Usage:\n"
    + "second line ...\n"
    + "third line line ...\n"
    + "last line");
```

Comment (quantumagi) [reply src/Stratis.Bitcoin.Cli/Program.cs]: I've removed the '$' signs.

Consistency first. The purpose of the StringBuilder is to remain consistent with the code found in the PrintHelp methods found in "RpcSettings.cs" and "MinerSettings.cs". If we are going to change it we should do it in those methods as well. However that  is not the aim of this PR.

Comment (bokobza) [reply src/Stratis.Bitcoin.Cli/Program.cs]: I'd rather use the StringBuilder here as it's easier to write than to use \n.

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 5]
```diff
@@ -10,53 +14,136 @@ namespace Stratis.Bitcoin.Cli
...
+#if DEBUG
+            isDebugMode = true;
+#endif
+            Exception ex = exception;
+            StringBuilder stringBuilder = new StringBuilder(128);
+            while (ex != null)
+            {
+                if (isDebugMode)
+                {
+                    stringBuilder.Append(ex.GetType().Name);
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Cli/Program.cs]: you can probably merge these 2 lines in one statement.

File: src/Stratis.Bitcoin.Cli/Program.cs [hunk 6]
```diff
@@ -10,53 +14,136 @@ namespace Stratis.Bitcoin.Cli
...
+                Console.WriteLine(ExceptionToString(err));
+            }
+        }
+
+        /// <summary>
+        /// Determine both the exception message and any inner exception messages.
+        /// </summary>
+        /// <param name="exception">The exception object.</param>
+        /// <returns>Returns the exception message plus any inner exceptions.</returns>
+        public static string ExceptionToString(Exception exception)
```
Comment (bokobza) [inline src/Stratis.Bitcoin.Cli/Program.cs]: Interesting, maybe we should put this as an extension method in the Utilities folder.
But we can do that in a later PR.

File: src/Stratis.Bitcoin.Cli/Program.cs (hunk 4 above)
Comment (Aprogiena) [reply src/Stratis.Bitcoin.Cli/Program.cs]: @quantumagi sure, no problem, i've just notice it here first, so just random comment, feel free to skip as it is off-topic

@bokobza then i think direct Console.WriteLine is better

File: src/Stratis.Bitcoin.Cli/Program.cs (hunk 5 above)
Comment (quantumagi) [reply src/Stratis.Bitcoin.Cli/Program.cs]: Done.

**Comment by bokobza** [comment]
Lovely, thank you.

**Pull request merged**